feat: automatically remove user registry secrets

olevski opened the pull request:
Closes #333 

A few things to note:
- ~~The scripts that are used by the cronjob can be moved to a fully independent docker image. I decided to not do this just because it was easier to piggy-back on the notebooks docker image.~~ This is now an independent image.
- ~~Use `{servername}-registry` to name user registry secrets (`servername` is escaped to prevent underscores from the project name causing issues.)~~ It was decided to use the following pattern `{safeusername}-registry-{uuid4()}`.
- Wrote unit tests following the ones that are already present for the api and other code
- The logic for removing secrets is as follows:
  1. In the cronjob check for any secret in the designated namespace that match the regex ~~`(.+)-registry$`~~ `.+-registry-[a-z0-9-]{36}$` and has the label `component=singleuser-server`, ~~with the capture group indicating the `servername` for the user server.~~
  2. In addition to the regex test also test for `secret.type == "kubernetes.io/dockerconfigjson"`and for the presence of the labels `["renku.io/commit-sha", "renku.io/projectName", "renku.io/username"]` which are used to then identify the pod that used the secret.
  3. If any secret passes all tests from the previous two steps, then:
    a) Check to see if the pod that used the secret is still around, if so and if the pod's status is running or completed, then delete the secret
    b) If the pod that used the secret cannot be found (based on the ~~servername~~ labels) then delete the secret only if the age of the secret is older than the designated threshold

This is currently in place at https://tasko.dev.renku.ch/. Tested with a public and private projects.

rokroskar commented:
I believe you need to also check the username of the pod explicitly - checking the server name is not enough. See https://github.com/jupyterhub/kubespawner/security/advisories/GHSA-v7m9-9497-p9gr - you can create user names that will map to the same escaped version of the name. Come to think of it, all of the secret handling probably needs to be rethought a bit - adding annotations would probably be the way to go.

ableuler reviewed (commented):
Thanks @olevski for this nicely done and well documented PR and particularly for test-covering the added functionality right away! 

 think the biggest change necessary is related to how we map a secret to a pod (see comment below). Particularly, there's a difference between the k8s pod name and the JH server name. While the k8s pod name IS unique and included both username, repo name and commit, different users can have servers with the same servername, even after fixing the CVE that @rokroskar has pointed out. For example check out https://<renku-domain>/jupyterhub/hub/home for a list of the server names of your running servers. My initial thought was to add the pod name directly as annotation to the secret, but the secret is created in the notebooks-service while the pod name is generated in the spawner, so we'd either have to duplicate the logic which generates the pod name or - what I have described in the comment - rely custom renku.io annotations (or labels!) for matching secrets and pods.

olevski commented on renku_notebooks/util/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,164 @@
+            f"Namespace {namespace} does not exist or the cluster is unreachable."
+        )
+    else:
+        for pod in pod_list.items:
```
@ableuler I decided to name the registry secrets using a `uuid4` from python and then appending `-registry` at the end. This way I have no way of accidentally overwriting a secret and I can always find the matching user JH pod based on the labels as you suggested. Let me know what you think.

olevski replied:
If you agree with this approach and all the other changes all I have to do is edit the tests to make sure they work with the new changes. This is an easy change though. But if I need to do other changes I would rather tackle those before finalizing the tests.

ableuler replied:
I think having the username still in the secret name is not bad in case you're looking for something manually. So maybe `{escaped_username}-{registry}-{random-uuid}?`

olevski replied:
@ableuler I fixed the secret names as you suggested. On the other related/side issues the status is as follows:
- The ui does indeed use some of the annotations we wanted to move as labels. Lorenzo confirmed this (see https://swiss-data-science.slack.com/archives/G67N59QL8/p1603372964031300). What Lorenzo suggested is that we can remap the k8s labels back into annotations when we return data from the API so this would not require changes to UI code at all and we would not have to have duplicates.
- The labels have the same key prefixes as the annotations
- I keep using the `safe_username` version of the username all throughout. I.e. both in labels and annotations, as well as in the pod and secret names (here we do not have a choice and we have to do this).

Let me know if you want me to do the label/annotation remapping as I mentioned in point 1. Otherwise this issue may be best tackled as a subtask of the larger task to implement schemas for renku-notebooks.

lorenzo-cavazzi replied:
If changes to the API require minor changes to UI (e.g. to keep the naming consistent and avoid confusion) it's fine.
Tbf I don't like the current object `annotation: { renku.io/<resource-name>: <value> }` structure, because the naming reflects the source name and any change makes it weird, like in this case where returning label data in an annotation object sounds confusing.

We can use generic names, like `repository` instead of `annotation`, and remove the `renku.io`  prefix. The `hub.jupyter.org/servername` could be moved one level up as `servername`.
```
repository: {
    namespace: <value>,
    commit: <value>,
    ...
},
servername: <value>
```

ableuler replied:
I think the question here is to what extent we want to just forward the K8s object, and how much we want to have "our own" resource that potentially deviates from the k8s resource. Following @lorenzo-cavazzi s suggestion would allow us to modify properties on the k8s object without changing the notebook-service API. On the other hand it's super obvious now where an annotation comes from which can make debugging errors maybe a bit easier? Either way, I think #421 is the proper place to discuss this.

olevski commented on renku_notebooks/util/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,164 @@
+    try:
+        secret_list = k8s_client.list_namespaced_secret(namespace)
+    except client.rest.ApiException:
+        logging.warning(
```
I removed all try/except/else blocks for the `k8s_client`

rokroskar commented:
Can we move this PR forward without making changes to the annotations/labels? And do that once #421 is done?

olevski commented:
> Can we move this PR forward without making changes to the annotations/labels? And do that once #421 is done?

Yes the last state of the code is to replicate the annotations as labels so that the UI does not break and the secrets are still automatically removed with the cron job.

rokroskar commented:
Ok great, that sounds good!

rokroskar commented on clean_user_registry_secrets/Pipfile:
```diff
@@ -0,0 +1,15 @@
+name = "pypi"
+
+[packages]
+escapism = "*"
+kubernetes = "*"
```
can we pin these to avoid surprises?

olevski replied:
I will pin them but btw they are not pinned in the pipenv file for the flask app. Should I pin there as well? There are quite a few packages that are not pinned there... I am not sure if you would like me to pin them all explicitly based on the lock file.

rokroskar replied:
yeah it's a bit sloppy there I agree. I suppose the lock file serves the purpose of "pinning" if we always strictly install with `--deploy` so maybe it's not necessary to pin the versions in the `Pipfile` itself?

ableuler replied:
Yes, with `pipenv` the idea is to only add known version restrictions to the `Pipfile` and leave the "pinning" to the lockfile.

olevski replied:
Oh ok cool. I will remove the pinning I did in the pipfile then.

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+
+def find_pod_by_secret(secret, k8s_client):
+    """Find the user jupyterhub podname based on the registry pull secret."""
+    label_keys = ['renku.io/commit-sha', 'renku.io/projectName', 'renku.io/username']
```
This shouldn't have passed the black formatting check so something looks broken in the tests/CI... but could you please format with black?

olevski replied:
yeah I saw that - I formatted the whole document and that line changed.

ableuler commented on clean_user_registry_secrets/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+
+LABEL maintainer="info@datascience.ch"
+
+RUN apk update && \
```
I think `apk update` is not needed when you use the `--no-cache` option for the install command.

ableuler reviewed (changes requested):
Looks really good. Just a few more minor changes from my side.

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+def remove_user_registry_secret(namespace, k8s_client, min_secret_age_hrs=0.25):
+    """Used in a cronjob to periodically remove old user registry secrets"""
+    secret_name_regex = ".+-registry-[a-z0-9-]{36}$"
+    label_keys = ['renku.io/commit-sha', 'renku.io/projectName', 'renku.io/username']
```
black formatting - see above

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+        if (
+            secret_name_match is not None
+            and secret.type == "kubernetes.io/dockerconfigjson"
+            and secret.metadata.labels.get("component") == "singleuser-server"
```
can't you use this label in a selector to reduce the number of secrets you loop through?

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+            secret_name_match is not None
+            and secret.type == "kubernetes.io/dockerconfigjson"
+            and secret.metadata.labels.get("component") == "singleuser-server"
+            and reduce(  # check that label keys for sha, project and username are present
```
maybe

```
all(label_key in secret.metadata.labels for label_key in label_keys)
```

is easier to understand? I'm pretty sure it would do the same thing but I could be wrong.

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+        # calculate secret age
+        tz = secret.metadata.creation_timestamp.tzinfo
+        secret_age = datetime.now(tz=tz) - secret.metadata.creation_timestamp
+        min_secret_age = timedelta(hours=min_secret_age_hrs)
```
doesn't this mean the max age a secret can have? Shouldn't it be called `max_secret_age` in that case? Also, this value doesn't change per secret so it shouldn't be set inside the loop

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+            podname = find_pod_by_secret(secret, k8s_client)
+            if podname is None:
+                # pod does not exist, delete if secret is old enough
+                if secret_age > min_secret_age:
```
if no pod using the secret exists, why do you care how old the secret is? To avoid the race condition where the culling process would catch a secret in between secret creation and pod creation? In that case the TTL of the secret can be much shorter imo

olevski replied:
@rokroskar yes I added this to prevent such a scenario. The secrets are fully unique now so there is no danger of rewriting something by mistake. So yeah my idea was better to have some secrets hang around for some time before they are deleted rather than delete them too quickly. 

As for you suggestion of using a TTL can you set a TTL on a secret? I don't see anything like that in documentation. Or maybe I misunderstand what you mean and you were suggesting to do something else?

olevski replied:
Oh I think I know what you mean - you are saying set the `-a` parameter (i.e. the max secret age) should be less than 15 minutes? I will set it to 5 minutes instead.

rokroskar replied:
yes that's what I meant - i don't see a scenario where it should me more than a few ms between the secret being created and the pod getting dispatched so even 5 minutes seems like a generous lifetime.

olevski replied:
@rokroskar I can go down to a lower value, 1 minute?

rokroskar replied:
haha no I think it's fine, I don't think it will cause any problems. If we see that the secrets are piling up too much for some reason we can reduce it further.

rokroskar commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+        "--age-hours-minimum",
+        type=float_gt_zero,
+        default=0.25,
+        help="The minimum age for a secret before it can be deleted if the user"
```
see above - I find it confusing to refer to this as the minimum age when it's the max age a secret is allowed to have :)

rokroskar reviewed (changes requested):
Thanks @olevski this looks good! Just a few minor comments for clarification.

olevski commented on chartpress.yaml:
```diff
@@ -9,6 +9,7 @@ charts:
       - .
       - jupyterhub
       - singleuser
+      - clean_user_registry_secrets
```
@rokroskar so I will just update the folder and image name but will keep the python file name and also function name as is.

olevski commented on clean_user_registry_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,151 @@
+# limitations under the License.
+"""Scripts used to remove user registry secrets in k8s"""
+
+from datetime import datetime, timedelta
```
@ableuler I assume that you still keep the python builtin packages vs the ones that are installed separate right? So because the only non-built-in package is kubernetes I only had to sort the group of built in packages.

rokroskar commented on cull_secrets/clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 - Swiss Data Science Center (SDSC)
```
```suggestion
# Copyright 2020 - Swiss Data Science Center (SDSC)
```

rokroskar commented on tests/test_clean_user_registry_secrets.py:
```diff
@@ -0,0 +1,242 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 - Swiss Data Science Center (SDSC)
```
```suggestion
# Copyright 2020 - Swiss Data Science Center (SDSC)
```

rokroskar reviewed (changes requested):
oops missed this the first time around.. copy/paste error I presume

olevski commented:
@ableuler , @rokroskar I addressed all the comments. Also I deployed the latest from this branch to my namespace and tested the cases when:
- the secret is deleted when the pod is around but is running
- the secret is not deleted when the age of the secret is less than 5 mins but the pod is not around
- the secret is deleted when the pod is not around and the secret is old enough

rokroskar commented:
One last thing and you might leave this to another PR, which would be fine imo, but we should have some way of controlling, through the values file, a) the cron setup and b) the max secret age

olevski commented:
> One last thing and you might leave this to another PR, which would be fine imo, but we should have some way of controlling, through the values file, a) the cron setup and b) the max secret age

No problem I can add this to this PR @rokroskar.

rokroskar reviewed (approved):
🚀
